### ISSUE TITLE
[BUG] Lombok incorrectly inherits constraints on sWerkplek01etter

### ISSUE THREAD

**Issue by HayoBaanAtHand** [projectlombok/lombok#3180]
**Describe the bug**

**To Reproduce**
Steps to reproduce the behavior, preferably in the form of the smallest java source file you can make that will show the problem when compiled with `javac -cp lombok.jar ExampleFile.java` or as sole java file in a fresh new eclipse project.

If this is not possible, give us enough to reproduce the problem. If you have stack traces or error messages please include all of them, and include screen shots if that will help explain the problem.

**Expected behavior**

**Version info (please complete the following information):**
 - Lombok version
 - Platform (javac or eclipse, and if so, what is the output of `javac -version` / the version listed in the _about..._ dialog of eclipse.

**Additional context**

**Comment by HayoBaanAtHand** [comment]
Yeah, sorry, accidentally pressed enter which saved the form I was editing. Completed it now..

**Comment by HayoBaanAtHand** [comment]
Note: the getters also incorrectly inherit the constraints.

**Comment by HayoBaanAtHand** [comment]
Thinking about this some more, I actually think you should not (at least not by default) inherit the annotations, not only because of above use case, but more in general: the validation is for the _end-state_ of the object, this does not necessarily say anything about the intermediate states. So the individual setters (and getters) should not be validated (again) as well.
There are use case where inheriting the annotations may be desired, of course, so adding an option that would do so would be nice still, just not by default and in general.

**Comment by rzwitserloot** [comment]
In all the hullabaloo I misunderstood a comment. I _thought_ us copying `validation` annotations has been part of lombok for quite a while already. I conflated the date of a PR with '... and it has been in lombok since that date' (whereas actually we accepted that PR only relatively recently).

In other words, I wasn't aware that 1.18.24 is the version that introduces this change.

Given the complexity / confused takes on what validation annotations even mean, "we wash our hands and do not do anything - we do not copy them, we do not add nullchecks" is probably better.

I'll talk to @rspilker about doing a quick 1.18.26 release to reduce the chance folks are going to start relying on the current copy behaviour. And I need to discuss this as a principle too, we may have to change a few things.

I'm going to make this the umbrella issue. The others are hopelessly bogged down and/or put the focus on an undesired way of addressing this problem.

**Comment by kflorian** [comment]
Any timeline for a "quick 1.18.26", as mentioned above?

**Comment by rPraml** [comment]
For search engine: Hibernate throws
```
javax.validation.ConstraintDeclarationException: HV000151:
A method overriding another method must not redefine the parameter constraint configuration,
but method MyClass#setFoo(String) redefines the configuration of MyInterface#setFoo(String).
```
example code:
```java
public interface MyInterface {
    void setFoo(String foo);
}

public class MyClass implements MyInterface {
    @Setter
    @javax.validation.constraints.NotNull
    private String foo;
}
```

As workaround, you may need to change the interface to `void setFoo(@NotNull String foo)` which is not always possible.
Waiting for 1.18.26...

**Comment by gitmotte** [comment]
Why copy annotations automatically without users knowlegde ... as far i've seen there is a solution for this already if someone likes to copy it ... 

```
@Getter(onMethod_ = { @NotNull })
@Setter(onParam_ = { @NotNull })
``` 

I don't think that Validation-Annotations should be copied at all - neither @NonNull nor other validations from the javax.validation namespace. The specification is clear and unambiguous - invalid data may be included in the bean - but not persisted by the JPA framework.

Apart from that, it is also possible to implement user-specific validations or to use an alternative extending validation framework, which would then not be copied as well.

I have to go back to version 1.8.22 until the change is rolled back!

**Comment by rzwitserloot** [comment]
> Why copy annotations automatically without users knowlegde

Because of the concept of type-associated annotations vs. field-associated ones.

The problem is, when annotations launched (softlaunch in JDK5, proper launch in JDK6), annotations were things you could stick on fields, methods, parameters, and types. Almost immediately people decided to use the feature to __annotate type usages__. The point of the annotation was that it was a modifier __of the type__, e.g. if the annotation was on a method, that really meant it was intended to be modifying the nature of the __return type__ of that method.

In a later release (I believe JDK8), this was enshrined by way of now allowing it everywhere, (i.e. the TYPE_USE target): Now, starting with jdk8, you can e.g. write `List<@NonNull String>`, which wasn't legal before.

The problem is, we still have a ton of legacy cruft:

* Annotations where the authors know perfectly well that it's properly targeted as TYPE_USE, but they launched them before TYPE_USE was out and do not consider it worth the backwards breakage to change them.
* Authors who don't understand the difference, even to this day,

and as a final kick in the behind, lombok's own limitations:

* Lombok can't _tell_ - we need resolution to know what the targets of an annotation are, and the JDK annotation spec is intentionally __unclear__ about what an annotation means if you don't have the annotation definition: `@Foo String bar;` can mean one of 2 completely different things (`@Foo` is an annotation on the type `String`, __or__ `@Foo` is an annotation on the field declaration `bar`). There's no way to know without checking the `@Target` annotation on the definition of `@Foo`.


For TYPE_USE annotations, it's _modifying the type_, therefore it stands to reason that if we say we __copy the type__ to e.g. the getter's return value, we copy that annotation just the same. But, when the annotation is supposed to apply to the field, it would be much more of a stretch if we copy them.

Point is: It is impossible to use the guidance of 'if on type, copy it, if on field, don't' for many reasons, but at the same time the semantic meaning of annotations means a blanket 'never ever copy it' is __also entirely wrong__ and it's not a fair solution to require from any lombok user to re-list all annotations in quadruplicate, telling lombok separately to copy them to the getter, the setter, the wither, the builder 'setter', the constructor parameter, etc.

So, we decided to auto-copy them.

Now, message is clear, this wasn't the right choice either. We've got alternate plans, which primarily involve:

* If it's type-level nullity associated, we know about it and copy appropriately and silently.
* Otherwise, if we recognize it in our list of known annotations, and that list says that copying MIGHT be warranted, we will generate a warning and demand that you state what you want us to do in your lombok.config. You can then choose to either have us apply the logical behaviour (presumably, copy it to various places), or not to (and once you have that setting, we stop generating a warning for it).
* It's backwards incompatible once, for a small set of annotations, namely the ones that aren't nullity related that we copy now. But they'd result in warnings that explain what you have to do, so that's not too bad.

> I have to go back to version 1.8.22 until the change is rolled back!

That is correct.

> I don't think that Validation-Annotations should be copied at all 

Note that this was a contribution. We didn't spend the appropriate time to evaluate it (a mistake we _rarely_ make, almost uniformly the complaint is that we don't accept enough PRs, not that we accept them too cavalierly!). We apologize for this.

**Comment by HayoBaanAtHand** [comment]
Any news on a fix release for this?

**Comment by Indifer** [comment]
@HayoBaanAtHand  You can go back to 1.8.22 first

**Comment by HayoBaanAtHand** [comment]
> @HayoBaanAtHand You can go back to 1.8.22 first

Sure, but I'd rather have prevented that as it means I have to add a version to the lombok dependency instead of being able to simply inherit the version as used by Spring Boot. Since I maintain 20+ microservices, I'd rather have prevented this and simply wait until this was fixed (and the fix assimilated in a Spring Boot update).

**Comment by kflorian** [comment]
> @HayoBaanAtHand You can go back to 1.8.22 first

No, I can't, because 1.8.22 doesn't support JDK 18.

**Comment by mjaggard** [comment]
`lombok.copyableAnnotations -= javax.validation.constraints.NotNull`
I've done this ^ which seems to have solved the problem for me (although I'm not totally clear whether that's the correct syntax or the problem has been solved by something else)

**Comment by mjaggard** [comment]
Looks that that syntax is correct (although I can't see it documented anywhere)
https://github.com/projectlombok/lombok/blob/master/src/core/lombok/core/configuration/ConfigurationParser.java#L101

**Comment by mjaggard** [comment]
Sorry, my mistake I'd removed the annotation from the field - it doesn't work.

**Comment by lildadou** [comment]
I encountered a problem with input validation in my Spring REST controllers. As the getters also have the @NotNull annotation then they are also checked during the object mapping. As a result, I get 2 constraint violations (one for the attribute and one for the method) and 2 strictly identical error messages are returned to my users.

**Comment by mjaggard** [comment]
More annoying than the bug in Lombok here is the fact that JPA doesn't check this until you actually use an entity - it's just asking for a bug in production 🤦‍♂️

**Comment by jamfor352** [comment]
Are there any updates on this?

**Comment by mjaggard** [comment]
@rzwitserloot do you think we can revert change #2904 or introduce a syntax for removing default copyable annotations?

**Comment by dstango** [comment]
Wondering if this issue can be closed due to https://github.com/projectlombok/lombok/commit/1473389f397579a224d57b496f6387f3b8cbb6a9 ...

**Comment by Hayo-Baan** [comment]
Certainly looks like it. If I read the comments in the change correctly, all validation annotations are now excluded. And this is indeed the correct approach here.

I don't see 1.18.26 as available yet, so I can't test to confirm, but will do as soon as it is and report back here.

**Comment by rfelgent** [comment]
@Hayo-Baan v1.18.26 is available since 3rd February 23
